spec/datadog/profiling/stack_recorder_spec.rb

```diff
@@ -545,6 +549,108 @@ def sample_allocation(obj)
             expect(heap_sample.values[:'heap-live-samples']).to eq(sample_rate * heap_sample_every)
           end
         end
+
+        context 'on Rubies supporting rb_gc_force_recycle' do
+          before do
+            skip 'rb_gc_force_recycle is a no-op in current Ruby version' if RUBY_VERSION >= '3.1'
+            @recycled_sample_allocation_line = 0
+          end
+
+          def has_seen_id_flag(obj)
+            described_class::Testing._native_has_seen_id_flag(obj)
+          end
+
+          def create_obj_in_recycled_slot(should_sample_original: false)
+            obj = Object.new
+            sample_allocation(obj) if should_sample_original
+            @recycled_sample_allocation_line = __LINE__ - 1
+
+            # Get the id of the object we're about to recycle
+            obj_id = obj.object_id
+
+            # Force recycle the given object
+            described_class::Testing._native_gc_force_recycle(obj)
+
+            # Repeatedly allocate objects until we find one that resolves to the same id
+            loop do
+              # Instead of doing this one at a time which would be slow given id2ref will
+              # raise on failure, allocate a ton of objects each time, increasing the
+              # probability of getting a hit on each iteration
+              1000.times { obj = Object.new }
+              begin
+                return ObjectSpace._id2ref(obj_id)
+              rescue RangeError # rubocop:disable Lint/SuppressedException
+              end
+            end
+          end
+
+          it 'enforces seen id flag on objects on recycled slots that get sampled' do
+            recycled_obj = create_obj_in_recycled_slot
+
+            expect(has_seen_id_flag(recycled_obj)).to be false
+
+            sample_allocation(recycled_obj)
+
+            expect(has_seen_id_flag(recycled_obj)).to be true
+          end
+
+          it 'enforces seen id flag on untracked objects that replace tracked recycled objects' do
+            recycled_obj = create_obj_in_recycled_slot(should_sample_original: true)
+
+            expect(has_seen_id_flag(recycled_obj)).to be false
+
+            serialize
+
+            expect(has_seen_id_flag(recycled_obj)).to be true
+          end
+
+          it 'correctly handles lifecycle of objects on recycled slots that get sampled' do
+            recycled_obj = create_obj_in_recycled_slot
+
+            sample_allocation(recycled_obj)
+            sample_line = __LINE__ - 1
+
+            recycled_heap_sample_matcher = lambda { |sample|
+              (sample.values[:'heap-live-samples'] || 0) > 0 && sample.locations.any? do |location|
+                location.lineno == sample_line && location.path == __FILE__
+              end
+            }
+
+            recycled_sample = heap_samples.find(&recycled_heap_sample_matcher)
+            expect(recycled_sample).not_to be nil
```

Comment:
Really minor: The objects we're using here come from `profiling/spec_helper.rb` (from `samples_from_pprof`). Since these kinds of checks are coming up often, it may be useful to add them has helpers for those classes, so we can simplify the assertions:

```ruby
recycled_sample = heap_samples.find { |s| s.values?(:heap-live-samples) && s.location?(line: sample_line, path: __FILE__) }
```

Reply:
Added this!